### PR TITLE
Enable the "Retriable and non-retriable pod failures for jobs" feature into beta

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

- Tracking issue: https://github.com/kubernetes/enhancements/issues/3329

#### Special notes for your reviewer:

It depends on the following PRs:
- https://github.com/kubernetes/kubernetes/pull/112360
- https://github.com/kubernetes/kubernetes/pull/113324
- https://github.com/kubernetes/kubernetes/pull/113304

#### Does this PR introduce a user-facing change?

```release-note
Enable the "Retriable and non-retriable pod failures for jobs" feature into beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

 ```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures 
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
Can you link all the PRs that this depends on?

**Comment by mimowo** [comment]
> Can you link all the PRs that this depends on?

Done

File: pkg/apis/batch/types.go
Comment (mimowo) [reply pkg/apis/batch/types.go]: Done

File: pkg/apis/core/types.go
Comment (mimowo) [reply pkg/apis/core/types.go]: Done

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (mimowo) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:928]: No, it is due to the fact that the file is generated with the default flags, and some roles are added behind the flag: https://github.com/kubernetes/kubernetes/blob/4d2128b5236ff3c1d621f6ceb6add16205ff3450/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go#L257

However, there was indeed a bug with permissions fixed last month. The new e2e test ""pods evicted from tainted nodes have pod disruption condition"  added in `tests.go` in this PR will prevent such a regression.

File: test/e2e/framework/pod/resource.go
Comment (mimowo) [reply test/e2e/framework/pod/resource.go:639]: I don't need to wait in the test where this function is used as the test checks that the pod is terminating already (has `deletionTimestamp != nil`. DisruptionTarget is added before the delete call.

File: test/e2e/framework/pod/resource.go
Comment (mimowo) [reply test/e2e/framework/pod/resource.go:544]: I'm not sure why. Should we retry in case of conflicts or wait until the pod is gone? I think we don't need to wait for the pod to be deleted as GC deletes it at some point anyway. Most of other tests (not using finalizers) do not wait actively for the pod to be gone.

Comment (mimowo) [reply test/e2e/framework/pod/resource.go]: I see, updated using `pod_client.Update` which does retries

**Comment by alculquicondor** [comment]
lgtm, overall, but we need to merge the other PRs first.

File: test/e2e/apps/job.go
Comment (alculquicondor) [inline test/e2e/apps/job.go:136]: since e2e tests are expensive to run, I'm fine with just having one test (the one above).

The important thing we are testing is that actions by kubelet can be matched with a job policy.

@aojea to comment whether is worth reducing the number of E2E tests. More through scenarios should be done at integration level for the job controller (which we did already).

File: test/e2e/node/taints.go
Comment (alculquicondor) [inline test/e2e/node/taints.go:349]: if this is tested at integration level already, I don't see much value on this test.

What you can do to add more value is to have a job failure policy that matches the condition. If you agree, I think we can still merge this today and leave the follow up for after code freeze and before test freeze.

File: test/e2e/scheduling/preemption.go
Comment (alculquicondor) [inline test/e2e/scheduling/preemption.go:324]: same note about value: adding job failure policy into the mix would be ideal

**Review by alculquicondor — COMMENTED** [review]
/approve
for scheduler and job controller

File: test/e2e/apps/job.go
Comment (aojea) [reply test/e2e/apps/job.go:136]: These tests seems to be very light, it seems to run only in 4 sec

> should allow to use the pod failure policy on exit code to fail the job early

E2E test are not only for testing the feature itself, also to see how it integrates with the rest of the system.

An example is the bug with the kubelet pod lifecycle, it was reported because someone was running job and its nature made the bug to be more exacerbated when using Jobs.

I would keep them, as I said , is adding only 4 secs

File: test/e2e/node/taints.go
Comment (aojea) [reply test/e2e/node/taints.go:349]: in doubt add both XD

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (mimowo) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:1649]: This is a patch to add the `DisruptionTarget` condition during preemption (https://github.com/kubernetes/kubernetes/blob/2b7b5245eaff4ed60d84df67e4351b04ad9d45ae/pkg/scheduler/framework/preemption/preemption.go#L369). There are other dedicated tests to ensure the condition is added. Here just making it pass with similar approach as for delete request.

File: test/e2e/scheduling/preemption.go
Comment (mimowo) [reply test/e2e/scheduling/preemption.go:324]: Possibly, if there is consensus to add it then I prefer to do decouple it from this PR and add before Test Freeze.

File: test/e2e/node/taints.go
Comment (mimowo) [reply test/e2e/node/taints.go:349]: As for the tests mixing job feature policy with pod disruption conditions I prefer to decouple them into a separate PR.

Comment (alculquicondor) [reply test/e2e/node/taints.go:349]: that is true... the permission issue by itself is enough justification for the test.

I think you can add a label that matches the feature gates, so that when the features are disabled, the tests don't run.

Debugging shouldn't be a problem if we have good logging. If adding the test motivates us to improve our logging, that's also a win.

**Review by soltysh — APPROVED** [review]
controller and controller-related tests lgtm
/lgtm
/approve

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (liggitt) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:340]: it's surprising these don't need to handle the patch/delete requests in any way

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml:943]: is there a way to group this with the existing update pods/status permission above? it makes it way easier to see this is not actually expanding the power of this controller

**Review by liggitt — COMMENTED** [review]
API constant change lgtm

a couple questions about the patch tests and one request on the default RBAC role if possible

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go
Comment (mimowo) [reply pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:773]: This a unit test and the logic to generate the patch content is beyond the taint manager code, which prepares and sends the request: https://github.com/kubernetes/kubernetes/blob/c288251818d9d39ffc10d8d133a5094288e73b3b/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L138. On the other hand preparing the apply request is the responsibility of the class so testing the patch would verify the logic. However, this is a concern for other tests in this file and it was already ticketed: https://github.com/kubernetes/kubernetes/issues/111612. If you think this is the right approach I would work on this as a follow up PR.

Also, for the context, we already have integration tests which verify the condition is added: https://github.com/kubernetes/kubernetes/blob/c288251818d9d39ffc10d8d133a5094288e73b3b/test/integration/node/lifecycle_test.go#L171. An e2e test for the path is also introduced in this PR.

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go
Comment (mimowo) [reply pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:330]: Same answer as below

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (mimowo) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:340]: This unit test focuses asserts on nominating the node for preemption. It does not assert the modifications done to the pod so it ignores the DELETE (and now also PATCH) action. 

However, a simple (yet more powerful alternative) is to index the potential victim pods in the fake client so that the test does not fail trying to patch non-existing pods (proposed in the new commit).

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (mimowo) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: There are two ways I guess:
1. add the "patch" permission to pods/status unconditionally since the "update" is already there
2. process the permissions in the test to group together the verbs based on equality of other fields in the `PolicyRule` structure.

I pushed a commit implementing 1. which seems pragmatic. I do not like 2 as it might not be trivial
and may require maintenance if new fields are added to the `PolicyRule` structure

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (mimowo) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go]: @alculquicondor @soltysh  PTAL

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go
Comment (alculquicondor) [reply pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:773]: The patch would be a slice of bytes. I'm fine not checking it, given that we have integration tests.

File: plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml
Comment (liggitt) [reply plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml]: that sounds good to me, thanks

**Comment by liggitt** [comment]
RBAC and API changes lgtm

will defer to area approvers for ack of patch test at https://github.com/kubernetes/kubernetes/pull/113360#discussion_r1017677492

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:336]: great, this seems better.

**Comment by liggitt** [comment]
adding to milestone per code-freeze exception granted at https://groups.google.com/g/kubernetes-sig-release/c/tZPjSWW_g30/m/_FzB7nrUAAAJ?utm_medium=email&utm_source=footer

/milestone v1.26

File: pkg/controller/nodelifecycle/scheduler/taint_manager_test.go
Comment (mimowo) [reply pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:773]: Ok, closed there related issue with a comment: https://github.com/kubernetes/kubernetes/issues/111612#issuecomment-1308788403

**Comment by leonardpahlke** [comment]
exception request approved: [email](https://groups.google.com/a/kubernetes.io/g/release-team/c/a2mtWz2lAbk/m/_FzB7nrUAAAJ)

Your updated deadline to make any changes to your PR is [18:00 PST Friday 11th November 2022](https://everytimezone.com/s/042f76f9).

/milestone v1.26

**Pull request merged**
